networkx/generators/random_graphs.py

```diff
@@ -1114,14 +1185,23 @@ def random_shell_graph(constructor, seed=None):
     seed : integer, random_state, or None (default)
         Indicator of random number generation state.
         See :ref:`Randomness<randomness>`.
+    create_using : Graph constructor, optional (default=nx.Graph)
+        Graph type to create. Graph instances are not supported.
+        Multigraph and directed types are not supported and raise a ``ValueError``.
 
     Examples
     --------
     >>> constructor = [(10, 20, 0.8), (20, 40, 0.8)]
     >>> G = nx.random_shell_graph(constructor)
 
     """
-    G = empty_graph(0)
+    G = empty_graph(0, create_using)
+    check_create_using(G, directed=False, multigraph=False)
+    if create_using is not None and not isinstance(create_using, type):
+        raise TypeError(
+            "random_shell_graph does not support graph instances for `create_using`. "
+            "Please use a graph type."
+        )
```

Comment:
Thanks for this!!


This `random_shell_graph` uses create_using in two ways -- one to make the new graph and another to make helper graphs which then are combined with the first using "union".  I'm hoping we can make this work by only using `create_using` to create the first graph. Then we avoid this special case where instances are not allowed in `create_using`.

Is it true that all the `gnm_random_graph`s could be `nx.Graph` type without affecting the result? I believe the `union` operation does what we want here, but if not, we could replace that with `G.add_nodes_from(g.nodes)` and `G.add_edges_from(g.edges)`. I believe that the code adding edges between shells should stay the same.

If we can use `nx.Graph` in `gnm_random_graph` then let's take out the `create_using` argument in line 1214 and remove this check that `create_using` is a type or is None.

Reply:
Another option might be to use `create_using=G.__class__` for the construction of the helper graphs so they have the same type as but are distinct instances from `G`. What do you think?

Reply:
That sounds like a good approach. If there are any second-hand effects from add_edge/node in the subclass, handling it this way should take care of that. Most of the time it won't matter but explicitly using `G.__class__` seems like it should get most of the troublesome cases. :)